go/libraries/doltcore/sqle/writer/prolly_table_writer.go

```diff
@@ -294,41 +292,63 @@ func (w *prollyTableWriter) IndexedAccess(i sql.IndexLookup) sql.IndexedTable {
 }
 
 // Reset puts the writer into a fresh state, updating the schema and index writers according to the newly given table.
-func (w *prollyTableWriter) Reset(ctx context.Context, sess *prollyWriteSession, tbl *doltdb.Table, sch schema.Schema) error {
-	sqlSch, err := sqlutil.FromDoltSchema("", w.tableName.Name, sch)
-	if err != nil {
-		return err
+func (w *prollyTableWriter) Reset(ctx *sql.Context, sess *prollyWriteSession, tbl *doltdb.Table, sch schema.Schema) error {
```

Comment:
This is really bad, we really don't want to couple session state management to correct writer behavior like this. If this needs to happen for correctness, then the control of information flow needs to be one way only. Either the session owns this information and the caches of it and is responsible for clearing them, or this writer only manages its own internal state on reset.

Reply:
I'm not sure I follow this. The writer schema state needs to live in the session to be cached between transactions. It's only used for writing, so it's going to be coupled to writers. It's exclusively a property of the table, so anywhere we try to write the table seems valid to check for the cached write state as long as the table is well-formed (its schema is consistent with its data). If we passed the latest table to `GetTableWriter` would that alleviate your concerns? Then there's no ambiguity maybe, we're disregarding whatever is in the existing write session, replacing it with the schema values for the given table.

Reply:
So follow up on this, it's preexisting coupling probably worse than we originally thought. I can't add table parameter to these methods without breaking correctness. Certain writers rely on using the tables in `prollyTableWriter.workingSet` rather than whatever is available from the calling scope. I don't really like the indirection, but I would at least need to rewrite all of the fulltext logic to make this not the case. It's possible it extends to other areas. All of the action happens in `GetTableWriter`/`Reset` and package cycles prevent me from moving the logic out of `dsess`. The places where we have the context to create writers is the only place to manage the caching lifecycle.

Reply:
Punt I guess. This is at least a bit more self-contained now.